File: src/test/java/me/testaccount666/serversystem/utils/VersionTest.java
package me.testaccount666.serversystem.utils;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class VersionTest {

    @Test
    void testVersionComparison() {
        var version1 = new Version("3.1.0");
        var version2 = new Version("3.2.0");
        var version3 = new Version("3.1.0");

        assertTrue(version2.compareTo(version1) > 0, "3.2.0 should be greater than 3.1.0");
        assertTrue(version1.compareTo(version2) < 0, "3.1.0 should be less than 3.2.0");
        assertEquals(0, version1.compareTo(version3), "3.1.0 should equal 3.1.0");
    }

    @Test
    void testVersionEquality() {
        var version1 = new Version("3.1.0");
        var version2 = new Version("3.1.0");
        var version3 = new Version("3.2.0");

        assertEquals(version1, version2, "Same versions should be equal");
        assertNotEquals(version1, version3, "Different versions should not be equal");
        assertEquals(version1.hashCode(), version2.hashCode(), "Same versions should have same hash code");
    }

    @Test
    void testVersionToString() {
        var version = new Version("3.1.0");
        assertEquals("3.1.0", version.toString(), "toString should return the version string");
        assertEquals("3.1.0", version.getVersion(), "getVersionString should return the version string");
    }

    @Test
    void testInvalidVersions() {
        assertThrows(IllegalArgumentException.class, () -> new Version(null), "Null version should throw exception");
        assertThrows(IllegalArgumentException.class, () -> new Version(""), "Empty version should throw exception");
        assertThrows(IllegalArgumentException.class, () -> new Version("   "), "Whitespace version should throw exception");
        assertThrows(IllegalArgumentException.class, () -> new Version("3.1.a"), "Non-numeric version should throw exception");
        assertThrows(IllegalArgumentException.class, () -> new Version("3..1"), "Double dots should throw exception");
    }

    @Test
    void testVersionComparisonWithDifferentLengths() {
        var version1 = new Version("3.1");
        var version2 = new Version("3.1.0");

        // 3.1 should be considered less than 3.1.0 (shorter versions are considered smaller)
        assertTrue(version1.compareTo(version2) < 0, "3.1 should be less than 3.1.0");
        assertTrue(version2.compareTo(version1) > 0, "3.1.0 should be greater than 3.1");
    }

    @Test
    void testNullComparison() {
        var version = new Version("3.1.0");
        assertThrows(NullPointerException.class, () -> version.compareTo(null), "Comparing to null should throw exception");
    }
}

File: src/test/kotlin/me/testaccount666/serversystem/utils/VersionTest.kt
package me.testaccount666.serversystem.utils

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * Unit tests for the Version class.
 *
 * Tests version parsing, comparison, and validation.
 */
class VersionTest {

    @Test
    fun `should create version with valid format`() {
        val version = Version("1.2.3")
        assertEquals("1.2.3", version.version)
    }

    @Test
    fun `should trim whitespace from version string`() {
        val version = Version("  1.2.3  ")
        assertEquals("1.2.3", version.version)
    }

    @ParameterizedTest
    @ValueSource(strings = ["", "   ", "1..2", "1.2.", ".1.2", "1.2.3a", "1.2-3"])
    fun `should reject invalid version string`(input: String) {
        assertThrows<IllegalArgumentException> { Version(input) }
    }

    @Test
    fun `should compare equal versions as equal`() {
        val version1 = Version("1.2.3")
        val version2 = Version("1.2.3")
        assertEquals(0, version1.compareTo(version2))
    }

    @Test
    fun `should identify newer major version`() {
        val version1 = Version("2.0.0")
        val version2 = Version("1.9.9")
        assertTrue(version1 > version2)
    }

    @Test
    fun `should identify older major version`() {
        val version1 = Version("1.9.9")
        val version2 = Version("2.0.0")
        assertTrue(version1 < version2)
    }

    @Test
    fun `should identify newer minor version`() {
        val version1 = Version("1.2.0")
        val version2 = Version("1.1.9")
        assertTrue(version1 > version2)
    }

    @Test
    fun `should identify newer patch version`() {
        val version1 = Version("1.2.3")
        val version2 = Version("1.2.2")
        assertTrue(version1 > version2)
    }

    @Test
    fun `should handle version with different segment counts`() {
        val version1 = Version("1.2")
        val version2 = Version("1.2.0")
        assertTrue(version1 < version2)
    }

    @Test
    fun `should handle version with single segment`() {
        val version1 = Version("2")
        val version2 = Version("1")
        assertTrue(version1 > version2)
    }

    @Test
    fun `should correctly use version in comparable scenarios`() {
        val versions = listOf(
            Version("1.2.3"),
            Version("1.0.0"),
            Version("2.1.0"),
            Version("1.2.2")
        )
        val sorted = versions.sorted()

        assertEquals("1.0.0", sorted[0].version)
        assertEquals("1.2.2", sorted[1].version)
        assertEquals("1.2.3", sorted[2].version)
        assertEquals("2.1.0", sorted[3].version)
    }

    @Test
    fun `should convert to string correctly`() {
        val version = Version("4.1.0")
        assertEquals("4.1.0", version.toString())
    }
}

